### PR TITLE
[Fix] Registration fix on mini-browsers which do not have cookies #487

### LINKED ISSUE #487: [bug] App fails on mini-browsers which do not have cookies
After days of debugging I nailed down the problem to this line:

https://github.com/openwisp/openwisp-wifi-login-pages/blob/d762b179a449bc3435ab53020115501200462f87/client/utils/validate-token.js#L20

We cannot rely on cookies for anything critical and we should always have a fallback.

I tested this change which allows the registration to avoid failing: https://github.com/openwisp/openwisp-wifi-login-pages/commit/1ea5e0ab2a402d43b893a03b625c1c80700b0ea8.

However, it introduces other issues because after that the verification process does not work properly, so it needs more work, but the though part should be over, now we only need to ensure everything keeps working (automated testing and manual testing) also after the changes.

### PR DESCRIPTION
**Pull request by codesankalp**
Closes #487

### CODE REVIEW

File: client/utils/validate-token.js [hunk 1]
```diff
@@ -17,13 +17,18 @@ const validateToken = async (
   language,
 ) => {
   const url = validateApiUrl(orgSlug);
-  const authToken = cookies.get(`${orgSlug}_auth_token`);
+  const authToken = userData.auth_token
+    ? userData.auth_token
+    : cookies.get(`${orgSlug}_auth_token`);
```
Comment (nemesisdesign) [inline client/utils/validate-token.js:22]: what about:

```javascript
// get auth token from redux state, otherwise try getting it from cookies
const authToken = userData.auth_token || cookies.get(`${orgSlug}_auth_token`);
```

File: server/controllers/validate-token-controller.js [hunk 2]
```diff
@@ -19,7 +19,7 @@ const validateToken = (req, res) => {
       const validateTokenUrl = reverse("validate_auth_token", getSlug(conf));
       const timeout = conf.timeout * 1000;
       let {token} = req.body;
-      if (req.body.session === "false") {
+      if (req.body.session === "false" && token.includes(".")) {
```
Comment (nemesisdesign) [inline server/controllers/validate-token-controller.js:22]: why is this needed?

File: client/utils/needs-verify.js [hunk 3]
```diff
@@ -9,15 +9,15 @@ const needsVerify = (method, user, settings) => {
...
+      (user.method === "mobile_phone" || user.radius_user_token !== "") &&
       user.is_verified === false &&
       settings.mobile_phone_verification
     );
   }
 
   if (method === "bank_card") {
     return Boolean(
-      user.method === "bank_card" &&
+      (user.method === "bank_card" || user.radius_user_token !== "") &&
```
Comment (nemesisdesign) [inline client/utils/needs-verify.js:20]: also here and above I wonder why is this needed?

**Review by nemesisdesign — COMMENTED** [review]
Good progress @codesankalp, see my comments below.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
During testing, I found issues:

- after successful verification, the captive portal login is opened in a new window
- when using mobile phone verification: the status page does not show up

sankalp-reg-without-cookies

File: server/controllers/validate-token-controller.js (hunk 2 above)
Comment (codesankalp) [reply server/controllers/validate-token-controller.js:22]: When the cookie is used then it was signed from the express backend but if we set the token directly then it must not be signed, So, skipping the step of signing token if it does not include `.`

File: client/utils/needs-verify.js (hunk 3 above)
Comment (codesankalp) [reply client/utils/needs-verify.js:20]: Routing and redirection is based on `needsVerify` function that's why I have to change this to take effect in the organization-wrapper component.

**Comment by nemesisdesign** [comment]
Superseded by https://github.com/openwisp/openwisp-wifi-login-pages/pull/489.
